stairway/src/main/java/bio/terra/stairway/impl/StairwayImpl.java

```diff
@@ -0,0 +1,652 @@
+   */
+  public StairwayImpl(StairwayBuilder builder) throws StairwayExecutionException {
+    this.maxParallelFlights =
+        (builder.getMaxParallelFlights() == null)
+            ? DEFAULT_MAX_PARALLEL_FLIGHTS
+            : builder.getMaxParallelFlights();
+
+    this.maxQueuedFlights =
+        (builder.getMaxQueuedFlights() == null)
+            ? DEFAULT_MAX_QUEUED_FLIGHTS
+            : (builder.getMaxQueuedFlights() < MIN_QUEUED_FLIGHTS
+                ? MIN_QUEUED_FLIGHTS
+                : builder.getMaxQueuedFlights());
+
+    this.exceptionSerializer =
+        (builder.getExceptionSerializer() == null)
+            ? new DefaultExceptionSerializer()
+            : builder.getExceptionSerializer();
+
+    this.stairwayName =
+        (builder.getStairwayName() == null)
+            ? "stairway" + UUID.randomUUID().toString()
+            : builder.getStairwayName();
+
+    this.queueManager = new WorkQueueManager(this, builder.getWorkQueue());
+
+    this.applicationContext = builder.getApplicationContext();
+    this.quietingDown = new AtomicBoolean();
+    this.hookWrapper = new HookWrapper(builder.getStairwayHooks());
+    this.completedFlightRetention = builder.getCompletedFlightRetention();
+    this.retentionCheckInterval =
+        (builder.getRetentionCheckInterval() == null)
+            ? Duration.ofDays(1)
+            : builder.getRetentionCheckInterval();
+  }
+
+  /**
+   * Second step of initialization
+   *
+   * @param dataSource database to be used to store Stairway data
+   * @param forceCleanStart true will drop any existing stairway data and purge the work queue.
+   *     Otherwise existing flights are recovered.
+   * @param migrateUpgrade true will run the migrate to upgrade the database
+   * @throws StairwayShutdownException stairway is shutdown and cannot initialize
+   * @throws DatabaseOperationException failures to perform recovery
+   * @throws MigrateException migration failures
+   * @throws StairwayException other Stairway exceptions
+   * @throws InterruptedException on thread shutdown
+   * @return list of Stairway instances recorded in the database
+   */
+  public List<String> initialize(
+      DataSource dataSource, boolean forceCleanStart, boolean migrateUpgrade)
+      throws StairwayShutdownException, DatabaseOperationException, MigrateException,
+          StairwayException, InterruptedException {
+
+    // If we have been shut down, do not restart.
+    if (isQuietingDown()) {
+      throw new StairwayShutdownException("Stairway is shut down and cannot be initialized");
+    }
+
+    stairwayInstanceDao = new StairwayInstanceDao(dataSource);
+    flightDao =
+        new FlightDao(
+            dataSource, stairwayInstanceDao, exceptionSerializer, hookWrapper, stairwayName);
+    control = new ControlImpl(dataSource, flightDao, stairwayInstanceDao);
+    flightSupport = new FlightSupportImpl(this);
+    flightFactory = new FlightFactory();
```

Comment:
is there a reason these are instantiated here instead of the constructor? being here implies to me they need the dataSource somehow.

Reply:
FlightSupport is gone. I made FlightFactory static.